presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -552,13 +553,20 @@ protected Object visitCoalesceExpression(CoalesceExpression node, Object context
                 return values.get(0);
             }
 
+            HashSet<Expression> visitExpression = new HashSet<>();
             List<Expression> expressions = values.stream()
                     .map(value -> toExpression(value, type))
+                    .filter(x -> !visitExpression.contains(x))
+                    .peek(visitExpression::add)
                     .collect(Collectors.toList());
 
             if (expressions.isEmpty()) {
                 return null;
             }
+
+            if (expressions.size() == 1 || expressions.get(0) instanceof Literal) {
```

Comment:
What if it is not `Literal`? Why can't we optimize then?

Reply:
For non `Literal` we cannot be sure would it return null or not , it might depend on the symbol reference. So we didn't optimize them